crates/unleash-edge-client-api/src/heartbeat.rs

```diff
@@ -0,0 +1,271 @@
+use axum::extract::{FromRef, Query, State};
+use axum::http::StatusCode;
+use axum::response::IntoResponse;
+use axum::routing::post;
+use axum::{Json, Router};
+use serde::Deserialize;
+use std::sync::Arc;
+use tracing::{debug, instrument};
+use ulid::Ulid;
+use unleash_edge_appstate::AppState;
+use unleash_edge_appstate::edge_token_extractor::{AuthState, AuthToken};
+use unleash_edge_http_client::UnleashClient;
+use unleash_edge_types::enterprise::{HeartbeatResponse, LicenseStateResponse};
+use utoipa::IntoParams;
+
+#[derive(Clone, Debug, Deserialize, IntoParams)]
+#[serde(rename_all = "camelCase")]
+struct ConnectionId {
+    #[param(value_type = String, example = "01J5X6F2Q9H8K4R1S2T3U4V5W6")]
+    #[serde(deserialize_with = "deserialize_ulid")]
+    pub connection_id: Ulid,
+}
+
+fn deserialize_ulid<'de, D>(deserializer: D) -> Result<Ulid, D::Error>
+where
+    D: serde::Deserializer<'de>,
+{
+    let s = String::deserialize(deserializer)?;
+    Ulid::from_string(&s).map_err(serde::de::Error::custom)
+}
+
+#[utoipa::path(
+    path = "/heartbeat",
+    post,
+    context_path = "/api/client/edge-licensing",
+    responses(
+        (status = 202, description = "Connection id was accepted and will be forwarded to Unleash Server soon", body = HeartbeatResponse),
+        (status = 503, description = "Upstream Edge instance was unable to license itself and so cannot license downstream instances"),
+    ),
+    security(
+        ("Authorization" = [])
+    )
+)]
+#[instrument(skip(app_state, edge_token))]
+async fn heartbeat(
+    State(app_state): State<HeartbeatState>,
+    AuthToken(edge_token): AuthToken,
+    Query(query_params): Query<ConnectionId>,
+) -> impl IntoResponse {
+    tokio::spawn(async move {
+        match app_state
+            .client
+            .send_heartbeat(&edge_token, &query_params.connection_id)
+            .await
+        {
+            Err(e) => {
+                tracing::info!("Unexpected error sending heartbeat: {}", e);
+            }
+            Ok(_) => {
+                debug!("Successfully forwarded heartbeat for downstream instance");
+            }
+        };
+    });
+
+    (
+        StatusCode::ACCEPTED,
+        Json(HeartbeatResponse {
+            edge_license_state: app_state.license_state,
+        }),
+    )
+}
+
+#[derive(Clone)]
+pub struct HeartbeatState {
+    license_state: LicenseStateResponse,
+    client: Arc<UnleashClient>,
+}
+
+impl FromRef<AppState> for HeartbeatState {
+    fn from_ref(app_state: &AppState) -> Self {
+        let client = app_state
+            .token_validator
+            .as_ref()
+            .as_ref()
+            .map(|v| v.unleash_client.clone())
+            .expect("Not running in Edge mode but enterprise Edge requires this for licensing");
```

Comment:
Not a huge fan of this, I think this is a hack. But without reworking the app state flow to have a client and/or build the token validator on the fly from a client, there's no other way to get a client right now. Leaving this because I think our app state is already too broad and this is the lesser of two evils but I'm open to thoughts here